test/utils/read-rows-resumption.ts

```diff
@@ -0,0 +1,182 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import {describe, it} from 'mocha';
+import {ChunkTransformer} from '../../src/chunktransformer';
+import {ReadRowsResumptionStrategy} from '../../src/utils/read-rows-resumption';
+import * as assert from 'assert';
+import {GoogleError} from 'google-gax';
+
+describe('Bigtable/Utils/ReadrowsResumptionStrategy', () => {
+  const tableName = 'fake-table-name';
+  [
+    {
+      name: 'should generate the right resumption request with no options each time',
+      expectedResumeRequest: {
+        rows: {
+          rowKeys: [],
+          rowRanges: [{}],
+        },
+        tableName,
+      },
+      options: {},
```

Comment:
Can you add a test that starts with a full table scan, but adds a lastSeenRowKey? 

Try to have the full table scan represented in multiple ways (both an empty row_set, and a rowset with an empty set of keys and ranges)

Reply:
Also, do you have a test for a resumption that results in an empty rowset (i.e. a completed request?)

Reply:
> Can you add a test that starts with a full table scan, but adds a lastSeenRowKey?
> 
> Try to have the full table scan represented in multiple ways (both an empty row_set, and a rowset with an empty set of keys and ranges)

I'm not sure what you mean by `lastSeenRowKey`. `nodejs-bigtable` uses `lastRowKey` and `lastScannedRowKey` in the chunk transformer. Did you mean either of those? Also, I think a full table scan is when the library attempts to get all the data from a table, but I think providing createreadstream with an empty row set or an empty set of keys and ranges will not do a full table scan. There was a bug fixed a while ago where the library was doing full table scans when an empty row set is provided which is not desirable because full table scans are really time consuming.

I think the tests you are mentioning describe the behaviour of the entire client library and are not a unit test on the resumption strategy, so perhaps if we clarify what those tests are then we can add them there?

Reply:
The first case is pretty tricky to implement. For the second case there is currently a line of code in `canResume` that says not to retry if `this.rowKeys.length === 0 && this.ranges.length === 0` and we can't remove this line of code because it addresses the second case you mentioned. For the first case however, the rows and ranges will be empty because the user requested a full table scan, but the client won't retry because this check is in place. To fix this, I could have an additional parameter in the constructor that is true if the user wants to do a full table scan and treat that as a special case adding a row range of `(lastKey, infinity)` to address this?

Reply:
Actually, the first case already works as you described. Please see the test case that says `retries a failed read`.